File: backend/modules/leads/csv_parser.py
"""CSV parsing + per-row validation for lead uploads.

Kept dependency-free (stdlib ``csv``) and pure so the same logic can be
exercised from tests without touching FastAPI or SQLAlchemy.
"""

from __future__ import annotations

import csv
import io
import re
from typing import Iterable


# RFC-ish; good enough for upload-time triage. We let pydantic's EmailStr
# tighten the screws at commit time.
_EMAIL_RE = re.compile(r"^[^@\s]+@[^@\s]+\.[^@\s]+$")

# Accept loose phone formatting (parens, dashes, spaces, leading +).
# We normalize by stripping everything except digits, then require 7-15
# digits (E.164 max is 15, anything under 7 is almost certainly garbage).
_PHONE_ALLOWED_RE = re.compile(r"^[+\d\s().\-]+$")


# Column synonyms — the FE/UX lets users drop any CSV with reasonable
# headers and we'll guess the mapping. Lowercase + stripped on both sides.
_COLUMN_SYNONYMS: dict[str, tuple[str, ...]] = {
    "name": (
        "name",
        "full name",
        "full_name",
        "contact",
        "contact name",
        "lead",
        "lead name",
    ),
    "email": ("email", "email address", "e-mail", "work email"),
    "phone": (
        "phone",
        "phone number",
        "mobile",
        "mobile number",
        "cell",
        "contact number",
        "tel",
        "telephone",
    ),
    "company": ("company", "organization", "organisation", "account", "employer"),
    "industry": ("industry", "vertical", "sector"),
    "location": ("location", "city", "country", "region"),
    "tags": ("tags", "labels"),
}


def normalize_phone(raw: str) -> str:
    """Return only the digits in ``raw``; ``""`` if nothing usable."""
    return re.sub(r"\D", "", raw or "")


def detect_columns(headers: list[str]) -> dict[str, str | None]:
    """Map our canonical column ids to the CSV's actual header names."""

    lowered = {h.strip().lower(): h for h in headers}
    mapping: dict[str, str | None] = {}
    for canonical, synonyms in _COLUMN_SYNONYMS.items():
        match: str | None = None
        for syn in synonyms:
            if syn in lowered:
                match = lowered[syn]
                break
        mapping[canonical] = match
    return mapping


def _value(row: dict[str, str], header: str | None) -> str:
    if not header:
        return ""
    val = row.get(header)
    if val is None:
        return ""
    return val.strip()


def parse_csv_text(text: str) -> tuple[list[dict[str, str]], list[str]]:
    """Return (rows, headers). Raises ``ValueError`` if the CSV is empty
    or has no header row."""

    if not text.strip():
        raise ValueError("CSV is empty")

    reader = csv.DictReader(io.StringIO(text))
    headers = reader.fieldnames or []
    if not headers:
        raise ValueError("CSV is missing a header row")

    rows: list[dict[str, str]] = []
    for row in reader:
        if all((v or "").strip() == "" for v in row.values()):
            continue
        rows.append(row)
    return rows, headers


def validate_row(
    *,
    row_number: int,
    raw: dict[str, str],
    columns: dict[str, str | None],
) -> dict:
    """Return a ``UploadParsedRow``-compatible dict for a single row."""

    name = _value(raw, columns["name"])
    email = _value(raw, columns["email"])
    phone = _value(raw, columns["phone"])
    company = _value(raw, columns["company"]) or None
    industry = _value(raw, columns["industry"]) or None
    location = _value(raw, columns["location"]) or None

    tags_raw = _value(raw, columns["tags"])
    tags = (
        [t.strip() for t in tags_raw.split(",") if t.strip()]
        if tags_raw
        else None
    )

    errors: list[str] = []

    if not name:
        errors.append("name is required")
    elif len(name) > 255:
        errors.append("name is too long (max 255 chars)")

    if not phone:
        errors.append("phone is required")
    else:
        if not _PHONE_ALLOWED_RE.match(phone):
            errors.append("phone contains invalid characters")
        normalized = normalize_phone(phone)
        if len(normalized) < 7:
            errors.append("phone is too short (need at least 7 digits)")
        elif len(normalized) > 15:
            errors.append("phone is too long (max 15 digits, E.164)")

    if email and not _EMAIL_RE.match(email):
        errors.append("email is not a valid address")

    # Stash any extra columns the user uploaded as custom_fields so we
    # don't lose data the schema doesn't model.
    known_headers = {h for h in columns.values() if h}
    custom_fields: dict[str, str] = {}
    for k, v in raw.items():
        if not k or k in known_headers:
            continue
        cleaned = (v or "").strip()
        if cleaned:
            custom_fields[k.strip()] = cleaned

    return {
        "row_number": row_number,
        "name": name or None,
        "email": email or None,
        "phone": phone or None,
        "company": company,
        "industry": industry,
        "location": location,
        "tags": tags,
        "custom_fields": custom_fields or None,
        "status": "invalid" if errors else "valid",
        "errors": errors,
    }


def annotate_duplicates(
    rows: Iterable[dict],
    *,
    existing_normalized_phones: set[str],
) -> list[dict]:
    """Mark rows duplicate-against-file or duplicate-against-db.

    Rows already classified ``invalid`` are left untouched.
    """

    seen_in_file: set[str] = set()
    out: list[dict] = []
    for row in rows:
        if row["status"] == "invalid":
            out.append(row)
            continue

        normalized = normalize_phone(row.get("phone") or "")
        if not normalized:
            out.append(row)
            continue

        if normalized in seen_in_file:
            row = {
                **row,
                "status": "duplicate",
                "errors": [*row["errors"], "duplicate phone in this CSV"],
            }
        elif normalized in existing_normalized_phones:
            row = {
                **row,
                "status": "duplicate",
                "errors": [
                    *row["errors"],
                    "phone already exists in your workspace",
                ],
            }
        else:
            seen_in_file.add(normalized)

        out.append(row)
    return out

File: frontend/src/lib/csvParser.ts
/**
 * Client-side CSV parsing and row-level validation for the lead importer.
 *
 * Mirrors the backend csv_parser.py logic so users see the same validation
 * errors before the file ever hits the API.
 */

import type { ParsedRow, ParsedRowStatus } from "@/types/lead";

// Re-export for consumers that import from here rather than types/lead
export type { ParsedRow, ParsedRowStatus };

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

const EMAIL_RE = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
const PHONE_ALLOWED_RE = /^[+\d\s().\-]+$/;

/** Maps canonical column names to common CSV header synonyms (lowercase). */
export const COLUMN_SYNONYMS: Record<string, string[]> = {
  display_name: [
    "display_name",
    "display name",
    "lead_name",
    "lead name",
    "alias",
  ],
  name: [
    "name",
    "full name",
    "full_name",
    "contact",
    "contact name",
    "lead",
  ],
  email: ["email", "email address", "e-mail", "work email"],
  phone: [
    "phone",
    "phone number",
    "mobile",
    "mobile number",
    "cell",
    "contact number",
    "tel",
    "telephone",
  ],
  company: [
    "company",
    "organization",
    "organisation",
    "account",
    "employer",
  ],
  industry: ["industry", "vertical", "sector"],
  location: ["location", "city", "country", "region"],
  tags: ["tags", "labels"],
};

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface ColumnMapping {
  display_name: string | null;
  name: string | null;
  email: string | null;
  phone: string | null;
  company: string | null;
  industry: string | null;
  location: string | null;
  tags: string | null;
  [key: string]: string | null;
}

export interface ValidationSummary {
  total: number;
  valid: number;
  invalid: number;
  duplicate: number;
  missingRequired: number;
  invalidEmails: number;
}

export interface ParseResult {
  rows: ParsedRow[];
  headers: string[];
  columns: ColumnMapping;
  summary: ValidationSummary;
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

export function normalizePhone(raw: string): string {
  return (raw ?? "").replace(/\D/g, "");
}

/** Map CSV headers → canonical column names via synonym lookup. */
export function detectColumns(headers: string[]): ColumnMapping {
  const lowered: Record<string, string> = {};
  for (const h of headers) {
    lowered[h.trim().toLowerCase()] = h;
  }

  const mapping: ColumnMapping = {
    display_name: null,
    name: null,
    email: null,
    phone: null,
    company: null,
    industry: null,
    location: null,
    tags: null,
  };

  for (const [canonical, synonyms] of Object.entries(COLUMN_SYNONYMS)) {
    for (const syn of synonyms) {
      if (syn in lowered) {
        mapping[canonical] = lowered[syn];
        break;
      }
    }
  }

  return mapping;
}

function getValue(
  row: Record<string, string>,
  header: string | null
): string {
  if (!header) return "";
  return (row[header] ?? "").trim();
}

function validateSingleRow(
  raw: Record<string, string>,
  columns: ColumnMapping,
  rowNumber: number
): ParsedRow {
  const display_name = getValue(raw, columns.display_name) || null;
  const name = getValue(raw, columns.name);
  const email = getValue(raw, columns.email);
  const phone = getValue(raw, columns.phone);
  const company = getValue(raw, columns.company) || null;
  const industry = getValue(raw, columns.industry) || null;
  const location = getValue(raw, columns.location) || null;
  const tagsRaw = getValue(raw, columns.tags);
  const tags = tagsRaw
    ? tagsRaw
        .split(",")
        .map((t) => t.trim())
        .filter(Boolean)
    : null;

  const errors: string[] = [];

  if (!name) {
    errors.push("name is required");
  } else if (name.length > 255) {
    errors.push("name is too long (max 255 chars)");
  }

  if (!phone) {
    errors.push("phone is required");
  } else if (!PHONE_ALLOWED_RE.test(phone)) {
    errors.push("phone contains invalid characters");
  } else {
    const digits = normalizePhone(phone);
    if (digits.length < 7) {
      errors.push("phone is too short (need at least 7 digits)");
    } else if (digits.length > 15) {
      errors.push("phone is too long (max 15 digits)");
    }
  }

  if (email && !EMAIL_RE.test(email)) {
    errors.push("email is not valid");
  }

  // Collect unknown columns as custom_fields.
  const knownHeaders = new Set(
    Object.values(columns).filter(Boolean) as string[]
  );
  const customFields: Record<string, string> = {};
  for (const [k, v] of Object.entries(raw)) {
    if (!k || knownHeaders.has(k)) continue;
    const cleaned = (v ?? "").trim();
    if (cleaned) customFields[k.trim()] = cleaned;
  }

  return {
    row_number: rowNumber,
    display_name,
    name: name || null,
    email: email || null,
    phone: phone || null,
    company,
    industry,
    location,
    tags,
    custom_fields: Object.keys(customFields).length ? customFields : null,
    status: errors.length > 0 ? "invalid" : "valid",
    errors,
  };
}

function annotateDuplicates(rows: ParsedRow[]): ParsedRow[] {
  const seen = new Set<string>();
  return rows.map((row) => {
    if (row.status === "invalid") return row;
    const normalized = normalizePhone(row.phone ?? "");
    if (!normalized) return row;
    if (seen.has(normalized)) {
      return {
        ...row,
        status: "duplicate" as ParsedRowStatus,
        errors: [...row.errors, "duplicate phone in this file"],
      };
    }
    seen.add(normalized);
    return row;
  });
}

// ---------------------------------------------------------------------------
// CSV text parser (RFC-4180)
// ---------------------------------------------------------------------------

function parseCSVLine(line: string): string[] {
  const result: string[] = [];
  let current = "";
  let inQuote = false;

  for (let i = 0; i < line.length; i++) {
    const ch = line[i];
    if (inQuote) {
      if (ch === '"') {
        if (line[i + 1] === '"') {
          current += '"';
          i++;
        } else {
          inQuote = false;
        }
      } else {
        current += ch;
      }
    } else {
      if (ch === '"') {
        inQuote = true;
      } else if (ch === ",") {
        result.push(current);
        current = "";
      } else {
        current += ch;
      }
    }
  }
  result.push(current);
  return result.map((v) => v.trim());
}

function parseCSVText(text: string): {
  rows: Record<string, string>[];
  headers: string[];
} {
  const lines = text.split(/\r?\n/);
  let headerLine = "";
  let dataStart = 0;

  for (let i = 0; i < lines.length; i++) {
    if (lines[i].trim()) {
      headerLine = lines[i];
      dataStart = i + 1;
      break;
    }
  }

  if (!headerLine) throw new Error("CSV is empty");

  const headers = parseCSVLine(headerLine);
  if (headers.length === 0 || headers.every((h) => !h)) {
    throw new Error("CSV is missing a header row");
  }

  const rows: Record<string, string>[] = [];
  for (let i = dataStart; i < lines.length; i++) {
    const line = lines[i];
    if (!line.trim()) continue;
    const values = parseCSVLine(line);
    const row: Record<string, string> = {};
    for (let j = 0; j < headers.length; j++) {
      row[headers[j]] = values[j] ?? "";
    }
    if (Object.values(row).every((v) => !v.trim())) continue;
    rows.push(row);
  }

  return { rows, headers };
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/** Full parse → validate → duplicate-annotate pipeline. */
export function processCSV(text: string): ParseResult {
  const { rows: rawRows, headers } = parseCSVText(text);
  const columns = detectColumns(headers);
  let rows = rawRows.map((raw, i) => validateSingleRow(raw, columns, i + 1));
  rows = annotateDuplicates(rows);

  const summary: ValidationSummary = {
    total: rows.length,
    valid: rows.filter((r) => r.status === "valid").length,
    invalid: rows.filter((r) => r.status === "invalid").length,
    duplicate: rows.filter((r) => r.status === "duplicate").length,
    missingRequired: rows.filter((r) =>
      r.errors.some((e) => e.includes("required"))
    ).length,
    invalidEmails: rows.filter((r) =>
      r.errors.some((e) => e.includes("email"))
    ).length,
  };

  return { rows, columns, headers, summary };
}

/** Serialize all rows to a downloadable validation-report CSV string. */
export function generateValidationReport(rows: ParsedRow[]): string {
  const escape = (v: string | number | null | undefined): string => {
    const s = String(v ?? "");
    if (/[",\n\r]/.test(s)) return `"${s.replace(/"/g, '""')}"`;
    return s;
  };

  const lines = [
    ["row", "name", "email", "phone", "company", "status", "errors"].join(","),
    ...rows.map((r) =>
      [
        r.row_number,
        escape(r.name),
        escape(r.email),
        escape(r.phone),
        escape(r.company),
        r.status,
        escape(r.errors.join("; ")),
      ].join(",")
    ),
  ];

  return lines.join("\n") + "\n";
}
